src/stable_marriage.rs

```diff
@@ -1,96 +1,234 @@
-            let woman = &mut women[i_woman];
-
-            if let Some(i_man_in_w_prefs) = woman.get_pref_idx(i_man) {
-                if let Some(i_otherman) = woman.cnt_match {
-                    men[i_otherman].cnt_match = None;
-                }
-                men[i_man].cnt_match = Some(i_woman);
-                woman.cnt_match = Some(i_man);
-                woman.prefs.truncate(i_man_in_w_prefs);
-            }
+        if self.current_is_perfect || !matcher.is_admissible(&self.data, item) {
+            return false;
+        }
+        if let Some(ref cnt_man) = self.current_match {
+            matcher.cmp_3way(&self.data, &cnt_man.data, item) == Ordering::Greater
+        } else {
+            true
         }
     }
+
+    fn replace_match(&mut self, man: Man<P>) -> Option<Man<P>> {
+        let mut old_match = Some(man);
+        std::mem::swap(&mut self.current_match, &mut old_match);
+        old_match
+    }
 }
 
 // Computes a stable matching between two lists of individuals.
 // See https://en.wikipedia.org/wiki/Stable_marriage_problem
 // This implements an extended version of the Gale-Shapley algorithm that allows for some
 // individuals to not rank every individual from the other list, in which case those two
 // individuals will never be matched together. In particular, the lists need not be the same size.
-//
-// @arg men: A list of preference rankings (most preferred first) of indices in the list `women`
-// @arg women: A list of preference rankings (most preferred first) of indices in the list `men`
-pub fn stable_marriage(
-    men: Vec<Vec<usize>>,
-    women: Vec<Vec<usize>>,
-) -> (Vec<Option<usize>>, Vec<Option<usize>>) {
-    let (mut men, mut women) = init_matching(men, women);
-
-    saturate_matching(&mut men, &mut women);
-
-    (extract_matching(men), extract_matching(women))
+// This algorithm favors men.
+// Returns matchings from the women's perspective, and unmatched men.
+// The order of women is preserved from the input list.
+pub fn stable_marriage<M, W, P: Matcher<Item = M, Target = W>, Q: Matcher<Item = W, Target = M>>(
+    men: Vec<M>,
+    women: Vec<W>,
+    men_matcher: &P,
+    women_matcher: &Q,
+) -> (Vec<(W, Option<M>)>, Vec<M>) {
+    let mut women = women
+        .into_iter()
+        .map(|item| Woman {
+            data: item,
+            current_match: None,
+            current_is_perfect: false,
+        })
+        .collect::<Vec<Woman<P>>>();
+
+    let mut no_longer_engageables = Vec::new();
+    'outer_loop: for item in men {
+        let mut man = Man {
+            data: item,
+            prefs: vec![],
+        };
+
+        if let Some(woman) = men_matcher.find_perfect_match(&man.data, &mut women) {
+            woman.current_is_perfect = true;
+            woman.replace_match(man);
+            continue;
+        }
+        man.prefs = men_matcher.compute_preference_list(&man.data, &women, women_matcher);
+
+        // Loop while we still hold an engageable man
```

Comment:
Would reword for “Loop while the man we hold is still engageable”, so as to avoid the confusion I fell in, which is thinking that we're going to loop over men here.